### ISSUE TITLE
Detected not working optional chaining on global variables / optional DOM APIs

### ISSUE THREAD

**Issue by HolgerJeromin** [eslint/eslint#14232]
**Please describe what the rule should do:**

eslint should warn me that a browser (without a secure context) can throw and the optional chaining will not save the day:

```js
// Throws an exception in runtime with insecure origin
> caches?.has('anything');
VM262:1 Uncaught ReferenceError: caches is not defined

// Runs without exception in runtime (with secure and insecure origin)
window.caches?.has('anything');
```

**What new ECMAScript feature does this rule relate to?**
Same for all new APIs which are on the global object.

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)

**Why should this rule be included in ESLint (instead of a plugin)?**
I dont know.
**Are you willing to submit a pull request to implement this rule?**
No, sorry.

**Comment by HolgerJeromin** [comment]
ref https://github.com/microsoft/TypeScript/issues/43295

**Comment by mdjermanovic** [comment]
Hi @HolgerJeromin, thanks for the issue!

> Same for all new APIs which are on the global object.

We can't really know what APIs should be considered as new, but I think that a generic rule that warns on `foo?.()` and `foo?.bar`, where `foo` is a global variable not declared in the same script, makes sense.

Such usage of `?.` looks like an intention to avoid the `typeof` check:

```js
if (typeof foo !== "undefined") {
  foo();
}
```

`typeof foo` is safe for undeclared globals, but `foo?.()` isn't.

**Comment by HolgerJeromin** [comment]
Thank you for your comment. 

> ...not declared in the same script,...

Very good point.

Perhaps we can detect also:
`if (foo !== undefined)`

**Comment by nzakas** [comment]
@mdjermanovic It seems like `no-undef` is the better solution to that problem?

**Comment by mdjermanovic** [comment]
> It seems like `no-undef` is the better solution to that problem?

It depends on whether optional globals are defined as globals in the user's configuration.

It makes sense to define them as globals to avoid `no-undef` errors and have them as resolved variables for other rules once the `typeof` check is done.

`caches` from the original example is defined as a [browser global](https://github.com/sindresorhus/globals/blob/926294653dc526c513dce3523b4675f63cc00e87/globals.json#L407), so `no-undef` doesn't report it.

**Comment by nzakas** [comment]
Gotcha. I’m not sold on the need for this being a core rule, though. It doesn’t really seem to rise to the level of must-have for us.

**Comment by nzakas** [comment]
We aren’t moving forward with this, so closing.
